File: fitness_engine/nutrition/micronutrients.py
"""
Micronutrient targets: fiber, fruit & vegetable intake.

Sources:
- Fiber: rippedbody.com__micros (14 g per 1,000 kcal)
- Fruit & veg intake tiers: rippedbody.com__micros
- Vegan supplement protocol: rippedbody.com__advice-for-vegans (Phase-2 hook)
"""
from __future__ import annotations

from ..models.nutrition import MicronutrientTargets


FIBER_G_PER_1000_KCAL = 14

FRUIT_VEG_TIERS = [
    # (max_calories, cups_fruit, cups_veg)
    (2000, 2, 2),
    (3000, 3, 3),
    (4000, 4, 4),
    (float("inf"), 4, 4),
]


def compute_micronutrients(target_calories: float) -> MicronutrientTargets:
    """
    Compute micronutrient targets based on calorie intake.

    - Fiber: 14 g per 1,000 kcal
    - Fruit & veg: tiered by calorie intake
    """
    fiber_g = FIBER_G_PER_1000_KCAL * target_calories / 1000

    cups_fruit = 2
    cups_veg = 2
    for max_cal, f, v in FRUIT_VEG_TIERS:
        if target_calories <= max_cal:
            cups_fruit = f
            cups_veg = v
            break

    notes = [
        f"Fiber target: {fiber_g:.0f} g ({FIBER_G_PER_1000_KCAL} g per 1,000 kcal)",
        f"Fruit: {cups_fruit} cups/day, Veg: {cups_veg} cups/day",
        "At-risk nutrients for dieters: calcium, zinc, magnesium, iron, vitamin D.",
        "Maintain dairy + red meat + sun exposure to avoid deficiencies.",
        "Vegan/vegetarian Phase-2: add B12, iron, zinc, calcium, omega-3, D3, creatine.",
    ]

    return MicronutrientTargets(
        fiber_g=round(fiber_g, 0),
        fruit_cups=cups_fruit,
        veg_cups=cups_veg,
        notes=notes,
    )


__all__ = [
    "FIBER_G_PER_1000_KCAL", "FRUIT_VEG_TIERS",
    "compute_micronutrients",
]

File: download/fitn/fitn_engine/lib/src/nutrition/micronutrients.dart
/// Micronutrient targets. See spec §4.3 step 7.
library;

import '../models/nutrition.dart';
import '../utils/round.dart';

/// Compute micronutrients.
///
/// - Fiber = 14 g × targetKcal / 1000
/// - Fruit/veg cups: 2/2 (≤2000 kcal), 3/3 (≤3000), 4/4 (>3000).
MicronutrientTargets computeMicronutrients({
  required double targetKcal,
}) {
  final notes = <String>[];
  final fiber = 14 * targetKcal / 1000;

  int fruitCups, vegCups;
  if (targetKcal <= 2000) {
    fruitCups = 2;
    vegCups = 2;
  } else if (targetKcal <= 3000) {
    fruitCups = 3;
    vegCups = 3;
  } else {
    fruitCups = 4;
    vegCups = 4;
  }

  notes.add('Fiber: 14 × ${targetKcal.round()} / 1000 = ${round1(fiber)} g');
  notes.add('Fruit/veg cups: $fruitCups/$vegCups');

  return MicronutrientTargets(
    fiberG: round1(fiber),
    fruitCups: fruitCups,
    vegCups: vegCups,
    notes: notes,
  );
}
